File: app/src/test/java/com/rafaeltmbr/stopwatch/infra/presentation/mappers/ViewTimeMapperTest.kt
package com.rafaeltmbr.stopwatch.infra.presentation.mappers

import com.rafaeltmbr.stopwatch.infra.presentation.entities.ViewTime
import com.rafaeltmbr.stopwatch.infra.presentation.mappers.impl.TimeMapper
import org.junit.Assert
import org.junit.Test

class ViewTimeMapperTest {
    @Test
    fun mapToViewTimeFractions_paddingMinutesAndSecondsWithZeros() {
        val mapper = TimeMapper()
        val seconds = 0.76f
        val expected = ViewTime(
            minutes = listOf("0", "0"),
            seconds = listOf("0", "0"),
            fraction = listOf("7", "6")
        )
        val viewTime = mapper.mapToViewTime(seconds)
        Assert.assertEquals(expected, viewTime)
    }

    @Test
    fun mapToViewTimeFractions_roundToClosestHundredthPart() {
        val mapper = TimeMapper()
        val seconds = 0.477f
        val expected = ViewTime(
            minutes = listOf("0", "0"),
            seconds = listOf("0", "0"),
            fraction = listOf("4", "8")
        )
        val viewTime = mapper.mapToViewTime(seconds)
        Assert.assertEquals(expected, viewTime)
    }

    @Test
    fun mapToViewTimeSeconds_paddingMinutesWithZeros() {
        val mapper = TimeMapper()
        val seconds = 21.52f
        val expected = ViewTime(
            minutes = listOf("0", "0"),
            seconds = listOf("2", "1"),
            fraction = listOf("5", "2")
        )
        val viewTime = mapper.mapToViewTime(seconds)
        Assert.assertEquals(expected, viewTime)
    }

    @Test
    fun mapToViewTimeMinutes_paddingTwoDigitsWithZeros() {
        val mapper = TimeMapper()
        val seconds = 107.12f
        val expected = ViewTime(
            minutes = listOf("0", "1"),
            seconds = listOf("4", "7"),
            fraction = listOf("1", "2")
        )
        val viewTime = mapper.mapToViewTime(seconds)
        Assert.assertEquals(expected, viewTime)
    }

    @Test
    fun mapToViewTimeMinutes_shouldNotTruncateMinutesIntoHours() {
        val mapper = TimeMapper()
        val seconds = 4107.43f
        val expected = ViewTime(
            minutes = listOf("6", "8"),
            seconds = listOf("2", "7"),
            fraction = listOf("4", "3")
        )
        val viewTime = mapper.mapToViewTime(seconds)
        Assert.assertEquals(expected, viewTime)
    }

    @Test
    fun mapToViewTimeNegativeTime_shouldSetTimeToZero() {
        val mapper = TimeMapper()
        val seconds = -4107.43f
        val expected = ViewTime(
            minutes = listOf("0", "0"),
            seconds = listOf("0", "0"),
            fraction = listOf("0", "0")
        )
        val viewTime = mapper.mapToViewTime(seconds)
        Assert.assertEquals(expected, viewTime)
    }
}

File: ios/StopwatchTests/platform/presentation/mappers/ViewTimeMapperTest.swift
import Testing
@testable import Stopwatch

struct ViewTimeTest {
    @Test func mapFractions_paddingMinutesAndSecondsWithZeros() {
        let mapper = ViewTimeMapperImpl()
        let milliseconds = 760
        let expected = ViewTime(
            minutes: ["0", "0"],
            seconds: ["0", "0"],
            fraction: ["7", "6"]
        )
        let viewTime = mapper.map(milliseconds)
        #expect(viewTime == expected)
    }
    
    @Test func mapFractions_mapToClosestHundredthPart() {
        let mapper = ViewTimeMapperImpl()
        let milliseconds = 477
        let expected = ViewTime(
            minutes: ["0", "0"],
            seconds: ["0", "0"],
            fraction: ["4", "8"]
        )
        let viewTime = mapper.map(milliseconds)
        #expect(viewTime == expected)
    }
    
    @Test func mapSeconds_paddingMinutesWithZeros() {
        let mapper = ViewTimeMapperImpl()
        let milliseconds = 21_520
        let expected = ViewTime(
            minutes: ["0", "0"],
            seconds: ["2", "1"],
            fraction: ["5", "2"]
        )
        let viewTime = mapper.map(milliseconds)
        #expect(viewTime == expected)
    }
    
    @Test func mapMinutes_paddingTwoDigitsWithZeros() {
        let mapper = ViewTimeMapperImpl()
        let milliseconds = 107_120
        let expected = ViewTime(
            minutes: ["0", "1"],
            seconds: ["4", "7"],
            fraction: ["1", "2"]
        )
        let viewTime = mapper.map(milliseconds)
        #expect(viewTime == expected)
    }
    
    @Test func mapMinutes_shouldNotTruncateMinutesIntoHours() {
        let mapper = ViewTimeMapperImpl()
        let milliseconds = 4_107_430
        let expected = ViewTime(
            minutes: ["6", "8"],
            seconds: ["2", "7"],
            fraction: ["4", "3"]
        )
        let viewTime = mapper.map(milliseconds)
        #expect(viewTime == expected)
    }
    
    @Test func map_shouldMapNegativeTimeToZero() {
        let mapper = ViewTimeMapperImpl()
        let milliseconds = -4_107_430
        let expected = ViewTime(
            minutes: ["0", "0"],
            seconds: ["0", "0"],
            fraction: ["0", "0"]
        )
        let viewTime = mapper.map(milliseconds)
        #expect(viewTime == expected)
    }
}
